### PR TITLE
Wip626082

### PR DESCRIPTION
**Pull request by BruceGoodGuy**
Hi @timhunt ,

This pull request contains 2 commits:

1. Limit the maximum and minimum size for cells.
I have upgraded my code to make sure the crossword is not too big or too small in some cases.
From now on, with the CellSize setting passed in from the renderer.php, we can decide the maximum and minimum value of the cell.

2. Fix inconsistencies on small devices.
With the first commit, I was causing an inconsistency when the user tried the crossword on mobile.
In the current behavior, based on the number of columns and rows (columns < 15 or rows < 15), the cell size will be the maximum size and the opposite of it will be the minimum size.
So in small devices, it is not suitable for user interface when the test has many crossword questions.
So in this fix, I force the cell size to always reach the min size on small devices.

### CODE REVIEW

File: amd/src/crossword_question.js [hunk 1]
```diff
@@ -446,4 +447,27 @@ export class CrosswordQuestion {
...
+    calculateCellSize(option) {
+        let size = option.cellSize[0];
+        option.isSmall = 0;
+        if (option.colsNum > 15 || option.rowsNum > 15 || window.innerWidth < 768) {
+            size = option.cellSize[1];
+            option.isSmall = 1;
+        }
+        // Calculate the size of the crossword: we need to add space between the cells.
+        // The size of the crossword is calculated by the formula:
+        // number of cells multiplied by (1 cell length + 1) and then 1.
```
Comment (timhunt) [inline amd/src/crossword_question.js:467]: No need to duplicate the code in the comment, so don't repeat the formula. The first line of the comment is what is useful. (You could make it clearer by changing it to something more precise like "The crossword size is the size of the cells plus the 1px borders."

What is not clear from the code, until you think about it, is that the +1 is becuase there is a 1px border between the cells. That is what the comment should say.

File: renderer.php [hunk 2]
```diff
@@ -47,7 +47,8 @@ public function formulation_and_controls(question_attempt $qa,
             'isPreview' => false,
             'title' => get_string('celltitle', 'qtype_crossword'),
             'orientation' => $orientationvalue,
-            'readonly' => false
+            'readonly' => false,
+            'cellSize' => [60, 40]
```
Comment (timhunt) [inline renderer.php:51]: We pass in cell size here, but then in lots of other places, we hard-code the magic numbers 60 and 40, so this is just a mess.

I could try to work out what is acutally going on, but acutally I won't. If it is only possible to understand the code after an hour of hard thought, the code is not good enough, and should first be rewritten to be much clearer, and then explained.

**Review by timhunt — CHANGES_REQUESTED** [review]
Please can you explain what is going on here.

I don't see any overall explanation of how the crossword is laid out, and that woudl help a lot.

However, what is very unhelpful is that the calculation of the layout seems to be spread between many different bits of code. It would be much clearer if all the code dealing with calculating the layout was in one or a few closely related methods.

Also, in the original issue, or this pull request, there is no explanation of what the change was. Before/after screen grabs might help, but words are required too.

Also, I am pretty sure what you have done is not what was requested. We still want this to be responsive, just with limits. You seem to have changed it to just use one of two sizes.

**Comment by BruceGoodGuy** [comment]
Sorry @timhunt ,
On previous commits, I misunderstood Chris' comment. And as per your comment, I changed the approach to make sure it supports responsiveness. Please refer to my new pull request:  https://github.com/moodleou/moodle-qtype_crossword/pull/11

And should I close this pull request? https://github.com/moodleou/moodle-qtype_crossword/pull/10

**Comment by timhunt** [comment]
Yes. I will close this.

**Pull request closed without merging**